# 통계 2일차

# 이항 분포

In [3]:
import numpy as np
from scipy import stats
import scipy.stats
import pandas as pd
import statsmodels.api as sm
import matplotlib.pyplot as plt
import seaborn as sns
from statsmodels.stats.proportion import proportions_ztest

In [5]:
# 이항 분포 연습
# n의 수
n = 3
for i in range(n+1):
    prob = stats.binom.pmf(k=i, n=n, p=0.4)
    print("P(X={0}) = {1:.3f}".format(i, prob))

P(X=0) = 0.216
P(X=1) = 0.432
P(X=2) = 0.288
P(X=3) = 0.064


In [6]:
# n의 수
n = 3
for i in range(n+1):
    prob = stats.binom.pmf(k=i, n=n, p=0.6)
    print("P(X={0}) = {1:.3f}".format(i, prob))

P(X=0) = 0.064
P(X=1) = 0.288
P(X=2) = 0.432
P(X=3) = 0.216


In [7]:
# 이항 분포 실습
# n의 수
n = 5
for i in range(n+1):
    prob = stats.binom.pmf(k=i, n=n, p=0.3)
    print("P(X={0}) = {1:.3f}".format(i, prob))

P(X=0) = 0.168
P(X=1) = 0.360
P(X=2) = 0.309
P(X=3) = 0.132
P(X=4) = 0.028
P(X=5) = 0.002


# 포아송 분포

In [27]:
# 평균
mu = 2
# 1분당 3번의 전화가 걸려올 확률
prob = stats.poisson.pmf(3, mu)
# 1분당 최대 2회 이하의 전화가 걸려올 확률
cdf_prob = stats.poisson.cdf(2, mu)
print("1분당 {0}번의 전화가 걸려올 확률: {1:.1f}%".format(3, prob * 100))
print("1분당 최대 {0}회 이하의 전화가 걸려올 확률: {1:.1f}%".format(2, cdf_prob * 100))
print("1분당 최대 {0}회 초과의 전화가 걸려올 확률: {1:.1f}%".format(2, (1-cdf_prob) * 100))

1분당 3번의 전화가 걸려올 확률: 18.0%
1분당 최대 2회 이하의 전화가 걸려올 확률: 67.7%
1분당 최대 2회 초과의 전화가 걸려올 확률: 32.3%


In [26]:
# 평균
mu = 3
# 1분당 3번의 전화가 걸려올 확률
prob = stats.poisson.pmf(2, mu)
# 1분당 최대 2회 이하의 전화가 걸려올 확률
cdf_prob = stats.poisson.cdf(1, mu)
print("{0}명의 손님이 도착 할 확률: {1:.1f}%".format(2, prob * 100))
print("{0}명 이하의 손님이 도착 할 확률: {1:.1f}%".format(1, cdf_prob * 100))
cdf_prob2 = stats.poisson.cdf(2, mu)
print("{0}명 이상의 손님이 도착 할 확률: {1:.1f}%".format(3, (1-cdf_prob2) * 100))

2명의 손님이 도착 할 확률: 22.4%
1명 이하의 손님이 도착 할 확률: 19.9%
3명 이상의 손님이 도착 할 확률: 57.7%


# 표본 분포

# 통계적 추정과 검정

# 정규성 검정


In [28]:
# 패키지 가져오기
from scipy.stats import shapiro
# camshaft데이터 가져오기
ds_camshaft = pd.read_csv("camshaft.csv", engine = "python")

In [29]:
ds_camshaft.head()

Length  Supp1  Supp2
0   601.4  598.0  601.6
1   601.6  599.8  600.4
2   598.0  600.0  598.4
3   601.4  599.8  600.0
4   599.4  600.0  596.8

In [31]:
data1 = ds_camshaft['Supp1']
data1

0     598.0
1     599.8
2     600.0
3     599.8
4     600.0
      ...  
95    599.6
96    599.8
97    599.0
98    599.6
99    599.4
Name: Supp1, Length: 100, dtype: float64

In [32]:
# 정규성 검정
shapiro(data1)

(0.9788156747817993, 0.10744144767522812)

샤프로-윌크 검정(Shapiro-Wilk test):
 - Shapiro-Wilk 검정통계량: 0.979
 - p-value :0.11 (정규성 있음)
   * 정규성 검정결과(95% 신뢰수준 사용시) p-value가 0.05보다 크면 정규, 0.05보다 작으면 비정규로 판단

In [33]:
# 정규성 검정, 실습
sample = pd.read_csv("2sample_t_test.csv", engine = "python")

In [39]:
sample.head()
print(sample)

    Mat_A  Mat_B  Total  Sub
0    73.4   68.7   73.4    1
1    77.0   71.4   77.0    1
2    73.7   69.8   73.7    1
3    73.3   75.3   73.3    1
4    73.1   71.3   73.1    1
5    71.5   72.7   71.5    1
6    74.5   66.9   74.5    1
7    77.5   70.2   77.5    1
8    76.4   74.4   76.4    1
9    77.7   70.1   77.7    1
10    NaN    NaN   68.7    2
11    NaN    NaN   71.4    2
12    NaN    NaN   69.8    2
13    NaN    NaN   75.3    2
14    NaN    NaN   71.3    2
15    NaN    NaN   72.7    2
16    NaN    NaN   66.9    2
17    NaN    NaN   70.2    2
18    NaN    NaN   74.4    2
19    NaN    NaN   70.1    2


In [42]:
mat_A = sample['Mat_A'].dropna(axis=0)
mat_B = sample['Mat_B'].dropna(axis=0)

In [46]:
print(mat_A)
print(mat_B)

0    73.4
1    77.0
2    73.7
3    73.3
4    73.1
5    71.5
6    74.5
7    77.5
8    76.4
9    77.7
Name: Mat_A, dtype: float64
0    68.7
1    71.4
2    69.8
3    75.3
4    71.3
5    72.7
6    66.9
7    70.2
8    74.4
9    70.1
Name: Mat_B, dtype: float64


In [44]:
shapiro(mat_A)

(0.9062636494636536, 0.2562994360923767)

샤프로-윌크 검정(Shapiro-Wilk test):
 - Shapiro-Wilk 검정통계량: 0.906
 - p-value :0.26 (정규성 있음)
   * 정규성 검정결과(95% 신뢰수준 사용시) p-value가 0.05보다 크면 정규, 0.05보다 작으면 비정규로 판단

In [45]:
shapiro(mat_B)

(0.9729753732681274, 0.9169777035713196)

샤프로-윌크 검정(Shapiro-Wilk test):
 - Shapiro-Wilk 검정통계량: 0.973
 - p-value :0.92 (정규성 있음)
   * 정규성 검정결과(95% 신뢰수준 사용시) p-value가 0.05보다 크면 정규, 0.05보다 작으면 비정규로 판단

# 모평균 추정

In [52]:
# 연습 모평균 추정

# df를 직접 입력
df = pd.DataFrame({"sample" : [3.6, 2.9, 2.8, 2.6, 3.4, 3.2, 2.2, 2.6, 2.6, 2.4,
                               2.4, 2.6, 3.4, 2.4, 2.1, 2.6, 3.0, 2.4, 2.7, 2.4,
                               2.6, 2.9, 2.9, 2.0, 2.7, 2.2, 2.8, 2.7, 1.8, 2.5,
                               3.0, 3.2, 2.8, 2.6, 3.2, 3.1, 2.9, 2.7, 2.7, 2.2]})

df.head()

sample
0     3.6
1     2.9
2     2.8
3     2.6
4     3.4

In [63]:
# df.variance()
a =np.mean(df) # 평균
b =np.std(df) # 분산
print(a)
print(b)

sample    2.695
dtype: float64
sample    0.388555
dtype: float64


In [60]:
print(a * 39/40)

sample    0.378841
dtype: float64


In [54]:
# 정규분포 구간 추정 함수 실행
lower, upper = stats.norm.interval(0.95, loc = np.mean(df), scale = 0.397/np.sqrt(40))
print("신뢰구간: {0},{1}".format(lower.round(2), upper.round(2)))

신뢰구간: [2.57],[2.82]


In [68]:
# 숙제 / 코딩 + 숙제
# 실습 모평균 구간 추정

df_prac = pd.DataFrame({"sample_prac": [18, 18, 20, 21, 20, 23, 19, 18, 17, 21,
                                         22, 20, 20, 21, 20, 19, 19, 18, 17, 19]})

lower_prac, upper_prac = stats.norm.interval(0.95, loc = np.mean(df_prac), scale = 3.8/np.sqrt(20))
print("신뢰구간: {0},{1}".format(lower_prac.round(2), upper_prac.round(2)))

신뢰구간: [17.83],[21.17]


$$x = {x bar -b \pm \sqrt{b^2-4ac} \over 2a}$$

$ \bar{X} \pm $  $ \sigma\over\sqrt{n} $

In [118]:
# 연습 모평균 추정 (모표준편차 모르는 경우)
# df 를 직접 입력
df = pd.DataFrame({"sample": [54.1, 53.3, 56.1, 55.7, 54.0, 54.1, 54.5, 57.1, 55.2, 53.8,
                              54.1, 54.1, 56.1, 55.0, 55.9, 56.0, 54.9, 54.3, 53.9, 55.0]})

# 정규분포 구간 추정 함수 실행
lower, upper = stats.t.interval(0.95, len(df)-1, loc = np.mean(df), scale = scipy.stats.sem(df))
print("신뢰구간: {0},{1}".format(lower.round(2), upper.round(2)))

신뢰구간: [54.39],[55.33]


In [111]:
# 연습 모평균 추정 (모표준편차 모르는 경우)
# df 를 직접 입력
df_pr = pd.DataFrame({"sample": [73, 71, 74, 69, 70, 73, 70, 68, 75, 72, 
                                 70, 72, 73, 70, 70, 72, 71, 70, 75, 72]})

# 정규분포 구간 추정 함수 실행
lower_pr, upper_pr = stats.t.interval(0.95, len(df_pr)-1, loc = np.mean(df_pr), scale = scipy.stats.sem(df_pr))
## scale = scipy.stats.sem(df_pr)은 s/sqrt(n)을 뜻함 
# the standard error of the mean 
print("신뢰구간: {0},{1}".format(lower_pr.round(2), upper_pr.round(2)))

신뢰구간: [70.6],[72.4]


In [112]:
np.mean(df_pr)

sample    71.5
dtype: float64

In [116]:
scipy.stats.sem(df_pr)

array([0.43225236])

In [115]:
scipy.stats.sem(df_pr, ddof = 1)

array([0.43225236])

In [75]:
np.mean(df_pr)

sample    71.5
dtype: float64

In [77]:
scipy.stats.sem(df_pr)**2

array([0.18684211])

In [80]:
np.sqrt(1.933)

1.3903237033151667

In [83]:
prop = stats.chi2.ppf(0.05, 9)
prop

3.325112843066815

In [84]:
prop = stats.chi2.ppf(0.025, 49)
prop

31.554916462667126

In [85]:
prop = stats.chi2.ppf(0.975, 49)
prop

70.22241356643451

In [89]:
df_call = pd.DataFrame({"sample": [12, 9, 10, 5, 16, 
                                 13, 5, 9, 8, 13, 
                                 16, 12, 12, 4, 14,
                                 9, 11, 14, 9, 17,
                                 5, 8, 9, 6, 11, 
                                 19, 6, 6, 8, 9,
                                 13, 10, 8, 13, 19,
                                 8, 12, 14, 16, 16,
                                 11, 17, 5, 9, 12,
                                 17, 11, 8, 5, 8]})


In [87]:
np.std(df_call)

sample    3.943653
dtype: float64

In [94]:
alpha = 0.05
n = len(df_call)
s = np.var(df_call['sample'], ddof = 1)
df = n -1

upper = (n-1) * s / stats.chi2.ppf(alpha/2, df)
lower = (n-1) * s / stats.chi2.ppf(1 - alpha/2, df)
print("신뢰구간: {0}, {1}".format(lower.round(2), upper.round(2)))

신뢰구간: 11.07, 24.64


In [96]:
# 연습 모비율 신뢰구간 추정
pro = stats.norm.ppf(0.05, 0, 1)
pro

-1.6448536269514729

In [104]:
p = 96 /1600
n = 1600
z = stats.norm.ppf(0.95, 0, 1)
print(p, n, pro)

0.06 1600 -1.6448536269514729


In [106]:
lower = p - z * np.sqrt(p*(1-p)/n)
upper = p + z * np.sqrt(p*(1-p)/n)

In [109]:
print("신뢰구간: ({0}, {1})".format(lower.round(3), upper.round(3)))

신뢰구간: (0.05, 0.07)


In [110]:
print(lower)

0.050234222675236584


In [125]:
# 실습 모비율 신뢰구간 추정
p = 9 /50
n = 50
z = stats.norm.ppf(0.975, 0, 1)
print(p, n, pro)

lower = p - z * np.sqrt(p*(1-p)/n)
upper = p + z * np.sqrt(p*(1-p)/n)
print("신뢰구간: ({0}, {1})".format(lower.round(3), upper.round(3)))

0.18 50 -1.6448536269514729
신뢰구간: (0.074, 0.286)


In [128]:
pro = stats.t.ppf(0.025, 9)
pro

-2.262157162740992

In [138]:
# 연습 평균 검정(모집단 표준편차를 모르는 경우)

# df를 직접 입력
df = pd.DataFrame({'sample': [74.5, 81.2, 73.8, 28.0, 76.3, 75.7, 80.2, 72.6, 77.9, 82.8]})

# 정규성 검정
shapiro(df)
print(shapiro(df))

(0.5709007978439331, 2.493129431968555e-05)


In [132]:
# 1-Sample t-test
t_result = stats.ttest_1samp(df, 76.7)
print(t_result)

Ttest_1sampResult(statistic=array([-0.87410318]), pvalue=array([0.4047826]))


In [136]:
# 1-Sample t-test의 t값 p값 저장
t, p = t_result.statistic.round(3), t_result.pvalue.round(3)
print(t, p)

[-0.874] [0.405]


In [137]:
# 1-Sample t-test의 t값 p값 출력
print("1-Sample t-test")
print("t 통계량 : {}".format(t))
print("p-value : {}".format(p))

1-Sample t-test
t 통계량 : [-0.874]
p-value : [0.405]


In [139]:
# 평균 검정(모집단 표준편차를 모르는 경우)

# df를 직접 입력
df_call = pd.DataFrame({'call': [85.0, 79.0, 79.1, 79.9, 21.6, 78.6, 85.4, 834, 78.1, 79.2]})

# 정규성 검정
shapiro(df_call)
print(shapiro(df_call))

(0.4223213195800781, 4.5214676447358215e-07)


In [140]:
# 1-Sample t-test
t_result = stats.ttest_1samp(df_call, 78.0)
print(t_result)

Ttest_1sampResult(statistic=array([0.94436839]), pvalue=array([0.36964162]))


In [141]:
# 1-Sample t-test의 t값 p값 저장
t, p = t_result.statistic.round(3), t_result.pvalue.round(3)
print(t, p)

[0.944] [0.37]


In [142]:
# 1-Sample t-test의 t값 p값 출력
print("1-Sample t-test")
print("t 통계량 : {}".format(t))
print("p-value : {}".format(p))

1-Sample t-test
t 통계량 : [0.944]
p-value : [0.37]
